pennylane/transforms/optimization/cancel_inverses.py

```diff
@@ -0,0 +1,144 @@
+    0: ──H──────H──────RZ(3)─────╭C──┤ ⟨Z⟩
+    1: ──H──────RY(2)──X──────X──│───┤
+    2: ──RX(1)──RX(2)────────────╰X──┤
+
+    We can see that there are two adjacent Hadamards on the first qubit that
+    should cancel each other out. Similarly, there are two Pauli-X gates on the
+    second qubit that should cancel. We can obtain a simplified circuit by running
+    the ``cancel_inverses`` transform:
+
+    >>> optimized_qfunc = cancel_inverses(qfunc)
+    >>> optimized_qnode = qml.QNode(optimized_qfunc, dev)
+    >>> print(qml.draw(optimized_qnode)(1, 2, 3))
+    0: ──RZ(3)─────────╭C──┤ ⟨Z⟩
+    1: ──H──────RY(2)──│───┤
+    2: ──RX(1)──RX(2)──╰X──┤
+    """
+    # Make a working copy of the list to traverse
+    list_copy = tape.operations.copy()
+
+    while len(list_copy) > 0:
+        current_gate = list_copy[0]
+        print(f"Working with current gate {current_gate}")
+        # Find the next gate that acts on at least one of the same wires
+        next_gate_idx = find_next_gate(current_gate.wires, list_copy[1:])
+
+        # If no such gate is found queue the operation and move on
+        if next_gate_idx is None:
+            apply(current_gate)
+            list_copy.pop(0)
+            continue
+
+        # Otherwise, get the next gate
+        next_gate = list_copy[next_gate_idx + 1]
+
+        # There are then three possibilities that may lead to inverse cancellation. For a gate U,
+        # 1. U is self-inverse, and the next gate is also U
+        # 2. The current gate is U.inv and the next gate is U
+        # 3. The current gate is U and the next gate is U.inv
+
+        # Case 1
+        are_self_inverses = current_gate.is_self_inverse and current_gate.name == next_gate.name
+
+        # Cases 2 and 3
+        are_inverses = False
+        name_set = set([current_gate.name, next_gate.name])
+        shortest_name = min(name_set)
+        if set([shortest_name, shortest_name + ".inv"]) == name_set:
+            are_inverses = True
+
+        # If either of the two flags is true, we can potentially cancel the gates
+        if are_self_inverses or are_inverses:
+            # If the wires are the same, then we can safely remove both
+            if current_gate.wires == next_gate.wires:
+                list_copy.pop(next_gate_idx + 1)
+            # If wires are not equal, there are two things that can happen
+            else:
+                # There is not full overlap in the wires; we cannot cancel
+                if len(Wires.shared_wires([current_gate.wires, next_gate.wires])) != len(
+                    current_gate.wires
+                ):
+                    apply(current_gate)
+                # There is full overlap, but the wires are in a different order
+                else:
+                    # If the wires are in a different order, gates that are "symmetric"
+                    # over all wires (e.g., CZ), can be cancelled.
+                    if current_gate.is_symmetric_over_all_wires:
```

Comment:
I've added a bit of extra logic here to handle the additional case where adjacent controlled operations share all controls and a target, but the controls may be in a different order. However this had led to extra branching, and this can probably be streamlined.